### DISCUSSION TITLE
Suspense with Async Component that relies on `searchParams`

### DISCUSSION THREAD

**Discussion by jbrtsnts** [vercel/next.js#50148] [App Router] [answered]
Hey everyone! 👋 I may have come across an issue related to Suspense behaviour with an Async Component that relies on `searchParams`.

https://codesandbox.io/p/sandbox/ecstatic-mirzakhani-s5hg9k

I've stumbled upon this problem and have created a sandbox to reproduce and analyse the issue. Feel free to explore the code and its behaviour, focusing on the following files within the sandbox:

- `/app/search/[keyword]/page.tsx`: This one I think works as expected, where the Suspense fallback is displayed while the `SearchResults` component fetches data.
  ```typescript
  export default function SearchPage({ params }: SearchPageProps) {
    const router = useRouter();
    return (
      <>
        <SearchForm
          value={params.keyword}
          onChange={(searchString) => router.replace(`/search/${searchString}`)}
        />
        <Suspense fallback={<p>Loading...</p>}>
          {/* @ts-expect-error Async Component */}
          <SearchResults queryString={params.keyword} />
        </Suspense>
      </>
    );
  }
   ```

- `/app/search/query/page.tsx`: this on the other hand ONLY works on the initial load as demonstrated [here](https://s5hg9k-3000.csb.app/search/query?q=laptop). And, when performing another search, Suspense no longer appears, but you can observe that it takes time to update the results as the data loads.
  ```typescript
  export default function SearchPage({ searchParams }: SearchPageProps) {
    const router = useRouter()
    return (
      <>
        <SearchForm
          value={searchParams.q}
          onChange={(searchString) => router.replace(`/search/query?q=${searchString}`)} />
        <Suspense fallback={<p>Loading...</p>}>
          {/* @ts-expect-error Async Component */}
          <SearchResults queryString={searchParams.q} />
        </Suspense>
      </>
    );
  }
   ```

What could be causing the difference in Suspense behaviour between the two components?
Are there any specific Next.js or React features that might be playing a role in this issue? 

Any help or guidance on what I might be missing or where I may have made a mistake would be greatly appreciated. Thank you!

**Comment by dmitkov28** [comment]
I'm facing the exact same issue. Adding key props to the Suspense component and its children seems to resolve it in dev mode, but the Suspense fallback is rendered inconsistently in prod.

**Comment by a95tbirdsc** [comment]
jbrtsnts,

Question for you.  If you removed your "<button>" from the code, could you get your example to work?  Meaning, you would start your search by entering in some text in your "<input>" field without clicking a button to search.  Like a Typeahead.  Query the database by what you are typing into the "<input>" field.

**Reply by jbrtsnts** [reply]
@a95tbirdsc - It's the same for [search as you type](https://s5hg9k-3000.csb.app/search/as-you-type?q=phone). What's happening is: it's fetching the data, the Suspense fallback never shows up, and the URL changes when the data is back.

I believe the expected behaviour would be the URL updates immediately, the Suspense fallback shows up, and then disappears once the data is fetched.

**Comment by kirchoni** [comment]
@jbrtsnts ,

I've faced similar issue with suspense fallback being shown only on initial render. In my exact case it turned out to be the `simulated` slowdown which interferes with how the suspense is being triggered.

I've tested my scenario with an external server which is _actually_ slow (or _artificially_ slowed, but on a separate process) and the fallback is shown as expected.

Hope you find a solution, but please let us know if removing the simulated slowdown and replacing it with an actually slow response interferes with the Suspense fallback in any way.

**Comment by jbrtsnts** [comment]
I tried a slightly different approach that _(kinda)_ works but it doesn't feel this is how it should be done 😄

https://s5hg9k-3000.csb.app/search/query-v2?q=laptop

#### Changes made
1. I wrapped with a client component `/app/search/query-v2/PageClientWrapper.tsx` which makes the `page.tsx` a server component. When the `page.tsx` have `'use client';` it still works but performs poorly - @kspeyanski, it feels like here's **simulated** slowdown plays a role.
2. Used the `useSearchParams()`, instead of `searchParams` from `page.tsx`
3. `SearchResults` is now **only** responsible for rendering results, I moved data `fetch` to the parent component. See 👇 

`PageClientWrapper.tsx`
```typescript
"use client";
import { useRouter, useSearchParams } from "next/navigation";
import { useState, Suspense } from "react";
import SearchResults2, { Product } from "./Results2";
import SearchForm from "../Form";
import asyncFetchProducts from "../asyncFetchProducts";

export default function PageClientWrapper() {
  const router = useRouter();
  const searchParams = useSearchParams();

  const queryString = searchParams.get("q") || "";
  const [products, seProducts] = useState<Product[]>();

  if (queryString) asyncFetchProducts(queryString).then(seProducts); // This is the change that feel made a difference;

  return (
    <>
      <SearchForm
        value={queryString}
        onChange={(searchString) =>
          router.replace(`/search/query-v2?q=${searchString}`)
        }
      />

      <Suspense fallback={<p>Loading...</p>}>
        {/* @ts-expect-error Async Component */}
        <SearchResults2 queryString={queryString} products={products} />
      </Suspense>
    </>
  );
}


```

The reason why I think this approach falls short is it still doesn't work with the reserve `loading.tsx` file.

**Reply by christianjuth** [reply]
In this example isn't `SearchResults2` a client component since you're importing it from a client component? 

I'm struggling to understand when to use Suspense. It seems it's most valuable when waiting for a Server Component (which requests/awaits data) to render. It's hard to understand the value here without seeing `SearchResults2`, but unless that component is very slow to render does suspending provide much benefit? Suspending during fetch makes more sense to me since some users will have slow internet leading to longer fetch times. Maybe I'm missing something obvious?

**Comment by zhanghaoda** [comment, marked as answer]
I've faced similar issue, and fix the problem when i add key prop to Suspense, like:

```
<Suspense key={suspenseKey} fallback={<Loading />}>
  {/*@ts-ignore*/}
  <DataTableComponent />
</Suspense>
```

suspenseKey could be the unique key relevant to the searchParams.

hope it helps!

**Comment by joonhyungshin** [comment]
I also had the same problem, and I believe @zhanghaoda 's solution is correct. There is also a relevant paragraph in the React doc:

https://react.dev/reference/react/Suspense#resetting-suspense-boundaries-on-navigation

**Comment by christianjuth** [comment]
This solution doesn't really make sense to me. Say I have these two files:
```jsx
// page.tsx

export default async function Page({ searchParams }) {
  const search results = fetchSearchResutls(searchParams)
  return (
    <>
      <FilterUI />
      <Suspense fallback={<span>...loading</span>}>
        <SearchResults key={"idk how to invalidate this since highlighted is client side state"} />
      </Suspense>
    </>
  )
}
```

```jsx
"use client"
export function FilterUI() {
  // Use state is the only way I've found to provide instant feedback when a filter is selected.
  // useSearchParams seems to not update until the server responds making it too slow on throttled network.
  const [highlighted, setHighlighted] = useState(0)
  return (
    <>
      <Link 
        href={{ query: { filter: 1 } }} 
        className={highlighted === 1 ? "bg-red" : ""}
        onClick={() => setHighlighted(1)}
      >
        Filter a
      </Link>
       <Link 
        href={{ query: { filter: 2 } }} 
        className={highlighted === 2 ? "bg-red" : ""}
        onClick={() => setHighlighted(2)}
      >
        Filter b
      </Link>
    </>
  )
}
```

I'm stuck in their weird loop where I need to update the UI instantly on slow network which requires a client component. I then need to invalidate the server component via a key. For this invalidation to not hang on slow network, I need to be able to invalidate it client side. But I don't think it's possible to get a client side state into a server component prop since any server component imported by a client component becomes a client component. Am I missing something obvious?

I just want a filter state that works on fast network and lets the user know something is happening (hiding outdated search results) on slow network.
